File: clients/csharp/src/Services/ObjectsService.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using MyMind.Actions;
using MyMind.Models;

namespace MyMind.Services;

public sealed class ObjectsService(MyMindClient client)
{
    public async Task<List<MindObject>> ListAsync(ListObjectsRequest? req = null, CancellationToken ct = default)
    {
        var q = new Dictionary<string, List<string>>();
        if (req?.Q is string s) q["q"] = [s];
        if (req?.Ids is { Count: > 0 } ids) q["id"] = ids;
        if (req?.ContentAs is string ca) q["contentAs"] = [ca];
        if (req?.Limit is int l) q["limit"] = [QueryFormat.Int(l)];

        var res = await client.SendAsync<ObjectListResult>(
            () => new HttpRequestMessage(HttpMethod.Get, $"/objects{QueryString.Build(q)}"), ct);
        return res.Objects ?? [];
    }

    public Task<MindObject> CreateAsync(CreateObjectRequest request, CancellationToken ct = default)
    {
        if (request.Blob is { } blob)
            return client.SendAsync<MindObject>(() => BuildMultipartCreate(request, blob), ct);

        return client.SendAsync<MindObject>(() => new HttpRequestMessage(HttpMethod.Post, "/objects")
        {
            Content = JsonContent.Create(request),
        }, ct);
    }

    public Task<MindObject> GetAsync(string id, string? contentAs = null, CancellationToken ct = default)
    {
        var q = contentAs is string ca
            ? new Dictionary<string, List<string>> { ["contentAs"] = [ca] }
            : null;
        return client.SendAsync<MindObject>(
            () => new HttpRequestMessage(HttpMethod.Get, $"/objects/{id}{QueryString.Build(q)}"), ct);
    }

    public Task<MindObject> UpdateAsync(string id, UpdateObjectRequest body, CancellationToken ct = default) =>
        client.SendAsync<MindObject>(() => new HttpRequestMessage(HttpMethod.Patch, $"/objects/{id}")
        {
            Content = JsonContent.Create(body),
        }, ct);

    public Task DeleteAsync(string id, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Delete, $"/objects/{id}"), ct);

    public Task RestoreAsync(string id, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Post, $"/objects/{id}/restore")
        {
            Content = JsonContent.Create(new { }),
        }, ct);

    public async Task<List<Match>> RelatedAsync(string id, int? limit = null, CancellationToken ct = default)
    {
        var q = limit is int l
            ? new Dictionary<string, List<string>> { ["limit"] = [QueryFormat.Int(l)] }
            : null;
        var res = await client.SendAsync<MatchListResult>(
            () => new HttpRequestMessage(HttpMethod.Get, $"/objects/{id}/related{QueryString.Build(q)}"), ct);
        return res.Matches ?? [];
    }

    public Task<Blob> DownloadAsync(string id, CancellationToken ct = default) =>
        client.SendAsync<Blob>(() => new HttpRequestMessage(HttpMethod.Get, $"/objects/{id}/download"), ct);

    public Task<string> GetContentAsync(string id, string accept, CancellationToken ct = default) =>
        client.SendAsync<string>(() =>
        {
            var req = new HttpRequestMessage(HttpMethod.Get, $"/objects/{id}/content");
            req.Headers.TryAddWithoutValidation("Accept", accept);
            return req;
        }, ct);

    public Task UpdateContentAsync(string id, string body, string contentType, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Put, $"/objects/{id}/content")
        {
            Content = new ByteArrayContent(Encoding.UTF8.GetBytes(body))
            {
                Headers = { ContentType = MediaTypeHeaderValue.Parse(contentType) },
            },
        }, ct);

    public Task PinAsync(string id, int? position = null, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Post, $"/objects/{id}/pin")
        {
            Content = JsonContent.Create(new PinRequest { Position = position }),
        }, ct);

    public Task UnpinAsync(string id, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Delete, $"/objects/{id}/pin"), ct);

    public Task TagAsync(string id, List<ObjectTag> tags, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Post, $"/objects/{id}/tags")
        {
            Content = JsonContent.Create(new AddObjectTagsRequest { Tags = tags }),
        }, ct);

    public Task AddToSpacesAsync(string id, List<ObjectSpace> spaces, CancellationToken ct = default) =>
        client.SendVoidAsync(() => new HttpRequestMessage(HttpMethod.Post, $"/objects/{id}/spaces")
        {
            Content = JsonContent.Create(spaces),
        }, ct);

    private static HttpRequestMessage BuildMultipartCreate(CreateObjectRequest request, Blob blob)
    {
        var multipart = new MultipartFormDataContent
        {
            { new StringContent(JsonSerializer.Serialize(request), Encoding.UTF8, "application/json"), "metadata" },
        };

        var blobContent = new StreamContent(blob.Stream);
        blobContent.Headers.ContentType = MediaTypeHeaderValue.Parse(blob.Type);
        if (blob.Name is string n) multipart.Add(blobContent, "blob", n);
        else multipart.Add(blobContent, "blob");

        return new HttpRequestMessage(HttpMethod.Post, "/objects") { Content = multipart };
    }
}

file sealed class ObjectListResult
{
    [JsonPropertyName("objects")]
    public List<MindObject>? Objects { get; init; }
}

file sealed class MatchListResult
{
    [JsonPropertyName("matches")]
    public List<Match>? Matches { get; init; }
}

File: clients/swift/src/Services/ObjectsService.swift
import Foundation

public final class ObjectsService: Sendable {
    private let client: ApiHttpClient

    init(client: ApiHttpClient) { self.client = client }

    public func list(_ req: ListObjectsRequest = .init()) async throws -> [MindObject] {
        var items: [URLQueryItem] = []
        if let q = req.q { items.append(.init(name: "q", value: q)) }
        if let ids = req.ids { ids.forEach { items.append(.init(name: "id", value: $0)) } }
        if let ca = req.contentAs { items.append(.init(name: "contentAs", value: ca)) }
        if let l = req.limit { items.append(.init(name: "limit", value: String(l))) }

        let res: ObjectListResult = try await client.send(.init(method: "GET", path: "/objects", queryItems: items))
        return res.objects ?? []
    }

    public func create(_ request: CreateObjectRequest) async throws -> MindObject {
        if let blob = request.blob {
            let (body, contentType) = try buildMultipartCreate(request: request, blob: blob)
            return try await client.send(.init(method: "POST", path: "/objects", body: body, contentType: contentType))
        }
        let body = try client.jsonBody(request)
        return try await client.send(.init(method: "POST", path: "/objects", body: body, contentType: "application/json"))
    }

    public func get(_ id: Uid, contentAs: String? = nil) async throws -> MindObject {
        var items: [URLQueryItem] = []
        if let ca = contentAs { items.append(.init(name: "contentAs", value: ca)) }
        return try await client.send(.init(method: "GET", path: "/objects/\(id)", queryItems: items))
    }

    public func update(_ id: Uid, _ req: UpdateObjectRequest) async throws -> MindObject {
        let body = try client.jsonBody(req)
        return try await client.send(.init(method: "PATCH", path: "/objects/\(id)", body: body, contentType: "application/json"))
    }

    public func delete(_ id: Uid) async throws {
        try await client.sendVoid(.init(method: "DELETE", path: "/objects/\(id)"))
    }

    public func restore(_ id: Uid) async throws {
        try await client.sendVoid(.init(method: "POST", path: "/objects/\(id)/restore", body: Data("{}".utf8), contentType: "application/json"))
    }

    public func related(_ id: Uid, limit: Int? = nil) async throws -> [Match] {
        var items: [URLQueryItem] = []
        if let l = limit { items.append(.init(name: "limit", value: String(l))) }
        let res: MatchListResult = try await client.send(.init(method: "GET", path: "/objects/\(id)/related", queryItems: items))
        return res.matches ?? []
    }

    public func download(_ id: Uid) async throws -> Blob {
        return try await client.download(.init(method: "GET", path: "/objects/\(id)/download"))
    }

    public func getContent(_ id: Uid, accept: String) async throws -> String {
        return try await client.send(.init(method: "GET", path: "/objects/\(id)/content", extraHeaders: ["Accept": accept]))
    }

    public func updateContent(_ id: Uid, body: String, contentType: String) async throws {
        let data = Data(body.utf8)
        try await client.sendVoid(.init(method: "PUT", path: "/objects/\(id)/content", body: data, contentType: contentType))
    }

    public func pin(_ id: Uid, position: Int? = nil) async throws {
        struct PinBody: Encodable { let position: Int? }
        let body = try client.jsonBody(PinBody(position: position))
        try await client.sendVoid(.init(method: "POST", path: "/objects/\(id)/pin", body: body, contentType: "application/json"))
    }

    public func unpin(_ id: Uid) async throws {
        try await client.sendVoid(.init(method: "DELETE", path: "/objects/\(id)/pin"))
    }

    public func tag(_ id: Uid, tags: [ObjectTag]) async throws {
        struct TagBody: Encodable { let tags: [ObjectTag] }
        let body = try client.jsonBody(TagBody(tags: tags))
        try await client.sendVoid(.init(method: "POST", path: "/objects/\(id)/tags", body: body, contentType: "application/json"))
    }

    public func addToSpaces(_ id: Uid, spaces: [ObjectSpace]) async throws {
        let body = try client.jsonBody(spaces)
        try await client.sendVoid(.init(method: "POST", path: "/objects/\(id)/spaces", body: body, contentType: "application/json"))
    }

    private func buildMultipartCreate(request: CreateObjectRequest, blob: Blob) throws -> (Data, String) {
        let boundary = UUID().uuidString
        let metadataJSON = try client.jsonBody(request)
        let blobName = blob.name ?? "blob"

        var body = Data()
        let boundaryLine = "--\(boundary)\r\n".data(using: .utf8)!

        body.append(boundaryLine)
        body.append("Content-Disposition: form-data; name=\"metadata\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: application/json\r\n\r\n".data(using: .utf8)!)
        body.append(metadataJSON)
        body.append("\r\n".data(using: .utf8)!)

        body.append(boundaryLine)
        body.append("Content-Disposition: form-data; name=\"blob\"; filename=\"\(blobName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(blob.type)\r\n\r\n".data(using: .utf8)!)
        body.append(blob.data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        return (body, "multipart/form-data; boundary=\(boundary)")
    }
}

// MARK: - Private response wrappers

private struct ObjectListResult: Decodable {
    let objects: [MindObject]?
}

private struct MatchListResult: Decodable {
    let matches: [Match]?
}
